Introduce a new platform layer type for Separated Images

etiennesegonzac opened the pull request:
#### 1842d1469e646940776d29bc3b89d5a60df7bd80
<pre>
Introduce a new platform layer type for Separated Images
<a href="https://bugs.webkit.org/show_bug.cgi?id=283721">https://bugs.webkit.org/show_bug.cgi?id=283721</a>
&lt;<a href="https://rdar.apple.com/137133500">rdar://137133500</a>&gt;

Reviewed by Simon Fraser.

Initially keyed on the `separated` transform style, and behind a
configuration flag, this patch introduces a new platformLayer Type
/ Custom View for Separated Images.

This view uses takes the layer&apos;s backing store IOSurface. And this
behavior is currently only enabled for images below the dimension
thresholds triggering tiling.

* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::isSeparatedImage const):
(WebCore::GraphicsLayer::setIsSeparatedImage):
New method to flag a graphics layer as a Separated Image.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setIsSeparatedImage):
Flag the graphics layer as a SepratedImage, will impact layer type.
(WebCore::GraphicsLayerCA::commitLayerTypeChangesBeforeSublayers):
Change the platform layer type for separated images.
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
Make sure Separated flag and clipping strategy are not lost when
changing layer type.

* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
Introduce new platform Layer Type for Separated Images.

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateSeparatedProperties):
Introduce a new method to update properties related to the Separated
transform style.
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateGeometry):
(WebCore::RenderLayerBacking::contentChanged):
Update Separated properties as part of the configuration update or when
the Image content changes.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerStyleChanged):
Flag the layer as needing a compositing configuration change when the
Separated transform-style is updated.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
WKA entry for the WKSeparatedImageView.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
Add a custom backing store application method for WKSeparatedImageViews,
passing the IOSurface to the view.

* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.h: Added.
* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.mm: Added.
(-[WKSeparatedImageView init]):
(-[WKSeparatedImageView initWithFrame:]):
(-[WKSeparatedImageView initWithCoder:]):
(-[WKSeparatedImageView setSurface:]):
Add a category to make sure this view is WKContentControlled regardless
of the implementation.
Stub implementation for public SDK builds.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Instantiate a custom view, WKSeparatedImageView for the Separated Image
Layer Type.

* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(allowListedClassToString):
(shouldDumpSeparatedDetails):
(dumpUIView):
Adjust the content of the text dump to get a reliable tests without
dynamic IDs. Dump layer corner radius for UIViews too.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-expected.txt: Added.
Add new UIView-based tests covering upgrading to and downgrading from a
Separated Image platform layer based on style, size and content changes.
They also cover cases where we don&apos;t use the `WKSeparatedImageView` for
layers using tiling.

Canonical link: <a href="https://commits.webkit.org/288372@main">https://commits.webkit.org/288372@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/27aee82cbaf34433c69ac55d8040486af5338159

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82891 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2540 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37185 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88008 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33945 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84984 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2614 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10457 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/64604 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/22365 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85947 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1940 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75405 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44878 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1839 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29594 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32979 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73066 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30276 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89374 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10181 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7401 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73021 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10409 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71211 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72236 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17927 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16397 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/15167 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/1562 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10134 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/15648 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10006 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13471 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -778,6 +778,15 @@ void GraphicsLayerCA::setIsSeparated(bool isSeparated)
+        return;
+
+    GraphicsLayer::setIsSeparatedImage(isSeparatedImage);
+    // Impacts layer type not properties.
```
I think you need to set some changed bit so that the commit code runs.

etiennesegonzac committed:
Introduce a new platform layer type for Separated Images
https://bugs.webkit.org/show_bug.cgi?id=283721
<rdar://137133500>

Reviewed by Simon Fraser.

Initially keyed on the `separated` transform style, and behind a
configuration flag, this patch introduces a new platformLayer Type
/ Custom View for Separated Images.

This view uses takes the layer's backing store IOSurface. And this
behavior is currently only enabled for images below the dimension
thresholds triggering tiling.

* Source/WebCore/platform/graphics/GraphicsLayer.cpp:
(WebCore::GraphicsLayer::GraphicsLayer):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::isSeparatedImage const):
(WebCore::GraphicsLayer::setIsSeparatedImage):
New method to flag a graphics layer as a Separated Image.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::setIsSeparatedImage):
Flag the graphics layer as a SepratedImage, will impact layer type.
(WebCore::GraphicsLayerCA::commitLayerTypeChangesBeforeSublayers):
Change the platform layer type for separated images.
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
Make sure Separated flag and clipping strategy are not lost when
changing layer type.

* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::operator<<):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::PlatformCALayerCocoa):
* Source/WebKit/Shared/Cocoa/WebCoreArgumentCodersCocoa.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::RemoteLayerTreeTransaction::description const):
Introduce new platform Layer Type for Separated Images.

* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateSeparatedProperties):
Introduce a new method to update properties related to the Separated
transform style.
(WebCore::RenderLayerBacking::updateConfiguration):
(WebCore::RenderLayerBacking::updateGeometry):
(WebCore::RenderLayerBacking::contentChanged):
Update Separated properties as part of the configuration update or when
the Image content changes.

* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::layerStyleChanged):
Flag the layer as needing a compositing configuration change when the
Separated transform-style is updated.

* Source/WebKit/Configurations/WebKit.xcconfig:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
WKA entry for the WKSeparatedImageView.

* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::drawInContext):
(WebKit::RemoteLayerBackingStoreProperties::applyBackingStoreToLayer):
Add a custom backing store application method for WKSeparatedImageViews,
passing the IOSurface to the view.

* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.h: Added.
* Source/WebKit/UIProcess/Cocoa/WKSeparatedImageView.mm: Added.
(-[WKSeparatedImageView init]):
(-[WKSeparatedImageView initWithFrame:]):
(-[WKSeparatedImageView initWithCoder:]):
(-[WKSeparatedImageView setSurface:]):
Add a category to make sure this view is WKContentControlled regardless
of the implementation.
Stub implementation for public SDK builds.
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteLayerTreeHost.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/UIProcess/RemoteLayerTree/ios/RemoteLayerTreeHostIOS.mm:
(WebKit::RemoteLayerTreeHost::makeNode):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
Instantiate a custom view, WKSeparatedImageView for the Separated Image
Layer Type.

* Source/WebKit/UIProcess/API/ios/WKWebViewTestingIOS.mm:
(allowListedClassToString):
(shouldDumpSeparatedDetails):
(dumpUIView):
Adjust the content of the text dump to get a reliable tests without
dynamic IDs. Dump layer corner radius for UIViews too.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-content.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-downgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled-expected.txt: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade-tiled.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-upgrade.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image.html: Added.
* LayoutTests/platform/visionos/transforms/separated-image-expected.txt: Added.
Add new UIView-based tests covering upgrading to and downgrading from a
Separated Image platform layer based on style, size and content changes.
They also cover cases where we don't use the `WKSeparatedImageView` for
layers using tiling.

Canonical link: https://commits.webkit.org/288372@main
